core/src/main/scala/sttp/tapir/server/interceptor/metrics/MetricsEndpointInterceptor.scala

```diff
@@ -0,0 +1,115 @@
+import sttp.tapir.server.interpreter.BodyListenerSyntax._
+import sttp.tapir.{DecodeResult, Endpoint}
+
+import scala.util.{Success, Try}
+
+class MetricsRequestInterceptor[F[_], B](metrics: List[Metric[F, _]], ignoreEndpoints: Seq[Endpoint[_, _, _, _]])
+    extends RequestInterceptor[F, B] {
+
+  override def apply(responder: Responder[F, B], requestHandler: EndpointInterceptor[F, B] => RequestHandler[F, B]): RequestHandler[F, B] =
+    new RequestHandler[F, B] {
+      override def apply(request: ServerRequest)(implicit monad: MonadError[F]): F[Option[ServerResponse[B]]] =
+        metrics
+          .foldLeft(monad.unit(List.empty[EndpointMetric[F]])) { (mAcc, metric) =>
+            for {
+              metrics <- mAcc
+              endpointMetric <- metric match {
+                case Metric(m, onRequest) => onRequest(request, m, monad)
+              }
+            } yield metrics :+ endpointMetric
+          }
+          .flatMap { endpointMetrics =>
+            requestHandler(new MetricsEndpointInterceptor[F, B](endpointMetrics, ignoreEndpoints)).apply(request)
+          }
+    }
+}
+
+private[metrics] class MetricsEndpointInterceptor[F[_], B](
+    endpointMetrics: List[EndpointMetric[F]],
+    ignoreEndpoints: Seq[Endpoint[_, _, _, _]]
+) extends EndpointInterceptor[F, B] {
+
+  override def apply(responder: Responder[F, B], endpointHandler: EndpointHandler[F, B]): EndpointHandler[F, B] =
+    new EndpointHandler[F, B] {
+      override def onDecodeSuccess[I](
+          ctx: DecodeSuccessContext[F, I]
+      )(implicit monad: MonadError[F], bodyListener: BodyListener[F, B]): F[ServerResponse[B]] = {
+        if (ignoreEndpoints.contains(ctx.endpoint)) endpointHandler.onDecodeSuccess(ctx)
+        else {
+          val responseWithMetrics: F[ServerResponse[B]] = for {
+            _ <- collectMetrics { case EndpointMetric(Some(onRequest), _, _) => onRequest(ctx.endpoint) }
+            response <- endpointHandler.onDecodeSuccess(ctx)
+            withMetrics <- withBodyOnComplete(response) {
+              case Success(_) => collectMetrics { case EndpointMetric(_, Some(onResponse), _) => onResponse(ctx.endpoint, response) }
+              case _          => ().unit
+            }
+          } yield withMetrics
+
+          responseWithMetrics.handleError { case e: Exception =>
+            collectMetrics { case EndpointMetric(_, _, Some(onException)) => onException(ctx.endpoint, e) }.flatMap(_ => monad.error(e))
+          }
+        }
+      }
+
+      /** If there's some `ServerResponse` collects `onResponse` as well as `onRequest` metric which was not collected in `onDecodeSuccess` stage.
+        */
+      override def onDecodeFailure(
+          ctx: DecodeFailureContext
+      )(implicit monad: MonadError[F], bodyListener: BodyListener[F, B]): F[Option[ServerResponse[B]]] = {
+        if (ignoreEndpoints.contains(ctx.endpoint)) endpointHandler.onDecodeFailure(ctx)
+        else {
+          ctx.failure match {
```

Comment:
I think we should attach run the metrics not based on the nature of the decode failure, but basing on its handling (we don't really know if a mismatch will be handled differently than other errors). So if downstream returns a response, then we need to run the on request & on response callbacks